src/pudl/transform/eia.py

```diff
@@ -583,82 +587,59 @@ def _boiler_generator_assn(  # noqa: C901
     Args:
         eia_transformed_dfs: a dictionary of post-transform dataframes
             representing the EIA database tables.
-        eia923_years: a list of the years of EIA 923 data that
-            should be used to infer the boiler-generator associations. By
-            default it is all the working years of data.
-        eia860_years: a list of the years of EIA 860 data that
-            should be used to infer the boiler-generator associations. By
-            default it is all the working years of data.
-        debug: If True, include columns in the returned dataframe
-            indicating by what method the individual boiler generator
-            associations were inferred.
 
     Returns:
-        eia_transformed_dfs (dict): Returns the same dictionary of dataframes
-        that was passed in, and adds a new dataframe to it representing
-        the boiler-generator associations as records containing
-        plant_id_eia, generator_id, boiler_id, and unit_id_pudl
+        A dataframe containing the boiler generator associations.
 
     Raises:
-        AssertionError: If the boiler - generator association graphs are not
-            bi-partite, meaning generators only connect to boilers, and boilers
-            only connect to generators.
-        AssertionError: If all the boilers do not end up with the same unit_id
-            each year.
-        AssertionError: If all the generators do not end up with the same
-            unit_id each year.
+        AssertionError: If the boiler - generator association graphs are not bi-partite,
+            meaning generators only connect to boilers, and boilers only connect to
+            generators.
+        AssertionError: If all the boilers do not end up with the same unit_id each
+            year.
+        AssertionError: If all the generators do not end up with the same unit_id each
+            year.
     """
+    debug = context.op_config["debug"]
+    eia_settings = context.resources.dataset_settings.eia
+    eia923_years = eia_settings.eia923.years
+    eia860_years = eia_settings.eia860.years
+
     if eia923_years is None:
         eia923_years = DataSource.from_id("eia923").working_partitions["years"]
     if eia860_years is None:
         eia860_years = DataSource.from_id("eia860").working_partitions["years"]
 
-    # if eia860_years or eia923_years are still empty, we can't compile the BGA.
-    # Return the unaltered input dictionary of dataframes instead.
     if not (eia860_years and eia923_years):
-        return eia_transformed_dfs
+        logger.warning("No EIA-860 or EIA-923 years specified. Returning input BGA.")
+        return eia_transformed_dfs["clean_boiler_generator_assn_eia860"]
+
+    # Do some final data formatting and assign appropriate types:
+    eia_transformed_dfs = {
+        table_name: convert_cols_dtypes(df, data_source="eia")
+        for table_name, df in eia_transformed_dfs.items()
+    }
```

Comment:
minor but since we are `_restrict_years`-ing on all of the tables, it might be nice to also do that here unless we are editing the input somehow (side note: are assets ever able to be edited after being created? as in do we have to worry about modifying the original?)

```suggestion
    # Do some final data formatting and assign appropriate types & restrict years:
    eia_transformed_dfs = {
        table_name: (
            convert_cols_dtypes(df, data_source="eia")
            .pipe(_restrict_years, eia923_years, eia860_years)
        )
        for table_name, df in eia_transformed_dfs.items()
    }
```

Reply:
We don't have to worry about editing the assets. Separate copies of whatever were generated upstream are passed in to each of the functions that operate on them downstream.

Do you mean do the `_restrict_years()` here **rather** than down below? That sounds like a good simplification if it should be done uniformly across all of the inputs!

Reply:
Incorporated `_restrict_years` into the little cleaning loop.